webapp/src/main/java/com/box/l10n/mojito/service/security/user/UserService.java

```diff
@@ -24,12 +29,43 @@ public class UserService {
   static Logger logger = getLogger(UserService.class);
 
   public static final String SYSTEM_USERNAME = "system";
+  private static final List<Role> ROLE_HIERARCHY = List.of(Role.ADMIN, Role.PM, Role.TRANSLATOR, Role.USER);
+  private static final Role MIN_ROLE = Role.PM;
   private final String rolePrefix = "ROLE_";
 
   @Autowired UserRepository userRepository;
 
   @Autowired AuthorityRepository authorityRepository;
 
+  @Autowired AuditorAwareImpl auditorAwareImpl;
+
+  private void requireMinRole(Role role) {
+    final Optional<User> currentUser = auditorAwareImpl.getCurrentAuditor();
+    if (currentUser.isEmpty()) {
+      // Can't happen in the webapp because only authenticated users may use
+      // the API endpoints. However, allow this for tests
+      return;
+    }
+    final String currentAuthority = currentUser.get().getAuthorities().iterator().next().getAuthority();
+    final Role currentRole = createRoleFromAuthority(currentAuthority);
+
+    if (ROLE_HIERARCHY.indexOf(currentRole) > ROLE_HIERARCHY.indexOf(role)) {
```

Comment:
> Also I am a bit curious why the Roles / Authorities are present in Mojito in the first place? Is this a Spring requirement or is there something already planed for the future?

For context, the original idea (7+ years ago) was to have authorization but we never got time to actually work on it on the server side. So, you can consider this as starting from scratch and if you find new ways in Spring to implement this, you should certainly consider them and make a proposal.

> Additionally, a [User can technically have multiple Authorities](https://github.com/box/mojito/blob/master/webapp/src/main/java/com/box/l10n/mojito/entity/security/user/User.java#L69-L71). However, practically each user has exactly one as far as I can tell

I think, we mapped to the Spring data model at the time in term of relationship cardinality. We implemented `UserDetails` interface backed by hibernate. Implementation is `UserDetailsImpl`. I see I eventually added `OAuth2UserDetailsImpl` (no handling of roles) but this points to `UserDetail` being a must have to have the overall spring security function properly. 

> Wouldn't it be possible to simplyfy this by using a @OneToOne relationship, or embedding the [Authority](https://github.com/box/mojito/blob/master/webapp/src/main/java/com/box/l10n/mojito/entity/security/user/Authority.java) directly into the user?

I think we need to keep the 1:N relationship. This was standard and probably still is. 

It would probably just replace the role hierarchy added in this PR. Eg. only the ROLE_ADMIN can create/modify any user user. ROLE_EDIT_USER can only CRUD on regular user/translator. 

I need to re-read all docs, that feel more like what I remember, vs having homegrown code to handle fine grain authorization.

Reply:
I have simplified the permission check.